File: Level-3 450 set  Pan-1/zero_sum_subarray.py

def findSubarray(arr):
    hashy = {}
    sum_to_find = 0
    res = 0
    sum_current = 0

    for i in range(len(arr)):
        # print(hashy)
        sum_current += arr[i]
        print("Sum current ", sum_current)
        # print("Sum to find", sum_to_find)
        if sum_current == sum_to_find:
            res += 1
        print("res", res)
        
        
        somt = sum_current - sum_to_find
        if somt in hashy:
            res += hashy[somt]
        if sum_current in hashy:
            hashy[sum_current] += 1
        else:
            hashy[sum_current] = 1
        print(hashy)

    print(res)



findSubarray([6,-1,-3,4,-2,2,4,6,-12,-7])
findSubarray([0,0,5,5,0,0])

File: Level-3 450 set  Pan-1/zero_sum_subarray.cpp
#include <iostream>
#include <bits/stdc++.h>
using namespace std;



int findSubarray(vector<int> arr, int n ) {
    unordered_map<int, int> sumSet;

    // with element zero 
    int sum_to_find = 0;

    int res = 0;
    int sum_current = 0;

    for(int i=0; i<n; i++){
        sum_current += arr[i];
        if (sum_current == sum_to_find){    
            res++;
        }

        if (sumSet.find(sum_current - sum_to_find) != sumSet.end()){
            res += (sumSet[sum_current - sum_to_find]);
        }

        sumSet[sum_current]++;
    }

    return res;

}




int main(){
    vector<int> arr = {6,-1,-3,4,-2,2,4,6,-12,-7};
    vector<int> arr2 = {0,0,5,5,0,0};

    cout << findSubarray(arr, 10);
    cout << findSubarray(arr2, 6);

    return 0;
}
